plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ExtraCredentialsBasedJdbcIdentityCacheMapping.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.plugin.jdbc.credential.ExtraCredentialConfig;
+
+import javax.inject.Inject;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class ExtraCredentialsBasedJdbcIdentityCacheMapping
+        implements JdbcIdentityCacheMapping
+{
+    private final Optional<String> userCredentialName;
+    private final Optional<String> passwordCredentialName;
+
+    @Inject
+    public ExtraCredentialsBasedJdbcIdentityCacheMapping(ExtraCredentialConfig config)
+    {
+        requireNonNull(config, "config is null");
+        userCredentialName = config.getUserCredentialName();
+        passwordCredentialName = config.getPasswordCredentialName();
+    }
+
+    @Override
+    public JdbcIdentityCacheKey getRemoteUserCacheKey(JdbcIdentity identity)
+    {
+        Map<String, String> extraCredentials = identity.getExtraCredentials();
+        return new ExtraCredentialsBasedJdbcIdentityCacheKey(
+                userCredentialName.map(extraCredentials::get),
+                passwordCredentialName.map(extraCredentials::get));
+    }
+
+    private static final class ExtraCredentialsBasedJdbcIdentityCacheKey
+            extends JdbcIdentityCacheKey
+    {
+        private final Optional<String> user;
+        private final Optional<String> password;
```

Comment:
Password as a cache key sounds weird, doesn't it?

Reply:
If we cached on user only we can easily make a mistake where we add a cache entry where proper password is provided. And then return data from cache even if some accesses Trino with wrong password.

Reply:
What you could maybe do here is to cache on sha256 of password. But I do not feel it is necessary.

Reply:
I'm also against keeping password here in raw form. Let's hash it (maybe altogether with username)?